### ISSUE TITLE
[Question] Helper class usage best practice

### ISSUE THREAD

**Issue by ValorHeart** [microsoft/playwright#18285]
Hello, What is the best practice for using helper classes with tests?
Given that the table interacting class constructor receives the page and the table locator (then inside it scan the table, etc...)

- Page object creates a new instance of the helper class and passes the page object and table locator into
- Test fixture creates a new instance of the helper class and passes the page object into, then the test itself passes the table locator into 'init' method, all of this without walking though the page object. 
- any other way?

Thanks!

**Comment by mxschmitt** [comment]
Depends totally on what kind of helper class it is, and what it requires. We at Playwright usually put things which are independent in small functions, some logic which belongs together we put into a class and make it available via fixtures if they depend on a page or other fixtures.

I guess you read the [advanced fixtures](https://playwright.dev/docs/test-fixtures) guide already, did you run into any issues while creating multiple fixtures? Or fixtures which depend on each other?

You can also create a helper class fixture and a page object fixture. Page object fixture uses the helper class fixture like any other fixture.

**Comment by ValorHeart** [comment]
@mxschmitt Thanks for the reply!
I don't really have an issue using fixtures, my issue is a theoretical practice that the Page Object is in charge of making the operations on the page, and the test 'talks' only with the page object rather than performing operations on the page straight.
my helper class is AgGrid table class that read the columns, rows, and rows cells and has methods to get all the required information; the constructor takes in page and table container locator.
1. if I'll put that helper class as a fixture, the test will need to be 'aware' of the table locator straight instead of the page object that the table belongs to so it will break the rule of POM.
2. to my thoughts, it's better than the test will call the page object and the page object will call the AgGrid table helper, the issue here is that the testers need to manually import the helper class to the page object class and won't have it optioned like a fixture.

What would you do in that case?

**Comment by mxschmitt** [comment]
I guess I'd import the helper class inside the POM, that sounds the most logical from what I can tell, but without seeing the code hard to reason about.

Have you tried it and does it work for you?

**Comment by ValorHeart** [comment]
@mxschmitt I did try it and it works, my main concern was how to make the developers aware of the existence of the helper class as they need to import it to the pom, I'm wrapping playwright with my own project, adding my config settings, fixtures, and helpers and then the development teams are importing it to their package.json.
with fixtures, they can see auto-complete options, but with stand-alone pom helpers, they could see it only at the import statement destructuring auto-complete; I guess it is good enough.
I did try to use the package.json conditional exports feature by attempting and allow them to import the helper class like "@scope/pw-toolkit/helpers" but I haven't managed to make it work. I saw that this playwright project is using that method of conditional exports as well, haven't managed to understand too well how it works (especially with typescript)

Thanks a lot for your time in the conversation!
